bolt-modules/boltlib/lib/puppet/functions/apply_prep.rb

```diff
@@ -8,155 +8,188 @@
+        )
+      end
+
+      inventory.add_facts(result.target, result.value)
     end
 
-    executor.log_action('install puppet and gather facts', targets) do
-      executor.without_default_logging do
-        # Skip targets that include the puppet-agent feature, as we know an agent will be available.
-        agent_targets, need_install_targets = targets.partition { |target| agent?(target, executor, inventory) }
-        agent_targets.each { |target| Puppet.debug "Puppet Agent feature declared for #{target.name}" }
-        unless need_install_targets.empty?
-          # lazy-load expensive gem code
-          require 'concurrent'
-          pool = Concurrent::ThreadPoolExecutor.new
-
-          hooks = need_install_targets.map do |t|
-            opts = t.plugin_hooks&.fetch('puppet_library').dup
-            plugin_name = opts.delete('plugin')
-            hook = inventory.plugins.get_hook(plugin_name, :puppet_library)
-            # Give plan function options precedence over inventory options
-            { 'target' => t,
-              'hook_proc' => hook.call(opts.merge(options), t, self) }
-          rescue StandardError => e
-            Bolt::Result.from_exception(t, e)
-          end
-
-          hook_errors, ok_hooks = hooks.partition { |h| h.is_a?(Bolt::Result) }
-
-          futures = ok_hooks.map do |hash|
-            Concurrent::Future.execute(executor: pool) do
-              hash['hook_proc'].call
-            end
-          end
-
-          results = futures.zip(ok_hooks).map do |f, hash|
-            f.value || Bolt::Result.from_exception(hash['target'], f.reason)
-          end
-          set = Bolt::ResultSet.new(results + hook_errors)
-          raise Bolt::RunFailure.new(set.error_set, 'apply_prep') unless set.ok
-
-          need_install_targets.each { |target| set_agent_feature(target) }
-        end
-
-        task = applicator.custom_facts_task
-        arguments = { 'plugins' => Puppet::Pops::Types::PSensitiveType::Sensitive.new(plugins) }
-        results = run_task(targets, task, arguments, options)
-
-        # TODO: Standardize RunFailure type with error above
-        raise Bolt::RunFailure.new(results, 'run_task', task.name) unless results.ok?
-
-        results.each do |result|
-          # Log a warning if the client version is < 6
-          if unsupported_puppet?(result['clientversion'])
-            Bolt::Logger.deprecate(
-              "unsupported_puppet",
-              "Detected unsupported Puppet agent version #{result['clientversion']} on target "\
-              "#{result.target}. Bolt supports Puppet agent 6.0.0 and higher."
-            )
-          end
-
-          inventory.add_facts(result.target, result.value)
-        end
+    results
+  end
+
+  # Return a list of targets and their puppet_library hooks.
+  #
+  private def get_hooks(targets, options)
+    hooks  = []
+    errors = []
+
+    targets.each do |target|
+      plugin_opts = target.plugin_hooks&.fetch('puppet_library').dup
```

Comment:
```suggestion
      plugin_opts = target.plugin_hooks.fetch('puppet_library').dup
```
We definitely don't handle this being nil, so...it must never be nil (and it does get set to empty hash in `inventory/target.rb`).